notes/trees-that-grow-proc-macro.md

```diff
@@ -0,0 +1,167 @@
+As nickel grows, it becomes unwieldy to use a single AST for every stage of compilation. Some things are kept out of the AST, even though they would be useful in some places, because they don't make sense and would get in the way at others (e.g. types for the LSP, closures at runtime). Other things are in the AST, but aren't actually filled in until later on in the process (e.g. forall's var_kind). This causes `unwrap()`s and `unreachable!()`s and hurts code maintainability.
```

Comment:
Nitpick, but Nickel is (not yet?) compiled :stuck_out_tongue: 

```suggestion
As Nickel grows, it becomes unwieldy to use a single abstract syntax tree (AST) for every stage of the pipeline. Some data are kept out of the AST, even though they would be useful in some places, because they don't make sense and would get in the way at others (e.g. types inferred during typechecking for the LSP, a closure - i.e. a term together with its environment - at runtime). Other data placeholder are in the AST, but aren't actually filled in until later on in the process (e.g. forall's var_kind). This causes `unwrap()`s and `unreachable!()`s and hurts code maintainability.
```